src/replication/replication_source_db_copy.cpp

```diff
@@ -0,0 +1,697 @@
+    release_stream ();
+
+    m_stream_file = NULL;
+    m_state = NOT_STARTED;
+    m_error_cnt = 0;
+  }
+
+  void source_copy_context::pack_and_add_object (multirow_object *&obj)
+  {
+    if (obj->get_rec_cnt () == 0)
+      {
+	delete obj;
+	obj = NULL;
+	return;
+      }
+
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::ACTIVE);
+
+    stream_entry.add_packable_entry (obj);
+    /* once packed, the object is owned by the stream entry */
+    obj = NULL;
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_statement (const std::string &str)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::ACTIVE);
+
+    sbr_repl_entry *sbr = new sbr_repl_entry (str.c_str (), "dba", "", NULL_LSA);
+    stream_entry.add_packable_entry (sbr);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_start_of_extract_heap (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::START_OF_EXTRACT_HEAP);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_end_of_extract_heap (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::END_OF_EXTRACT_HEAP);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_end_of_copy (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::END_OF_REPLICATION_COPY);
+
+    stream_entry.pack ();
+  }
+
+  int source_copy_context::execute_and_transit_phase (copy_stage new_state)
+  {
+    int error = NO_ERROR;
+
+    assert (new_state != m_state);
+
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::execute_and_transit_phase "
+			      "curr_state:%d, new_state:%d", m_state, new_state);
+
+    std::unique_lock<std::mutex> ulock_state (m_state_mutex);
+
+    if ((int) m_state != ((int) new_state - 1))
+      {
+	er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 0);
+	return ER_GENERIC_ERROR;
+      }
+
+    m_state = new_state;
+
+    /* some of the new states require specific actions */
+    if (new_state == SCHEMA_EXTRACT_CLASSES_FINISHED)
+      {
+	pack_and_add_statement (m_class_schema);
+      }
+    else if (new_state == SCHEMA_COPY_TRIGGERS)
+      {
+	pack_and_add_statement (m_triggers);
+      }
+    else if (new_state == SCHEMA_COPY_INDEXES)
+      {
+	pack_and_add_statement (m_indexes);
+      }
+
+    /* unlock after adding indexes to stream to */
+    ulock_state.unlock ();
+    m_state_cv.notify_one ();
+
+    return error;
+  }
+
+  int source_copy_context::wait_for_state (cubthread::entry &thread_ref, const copy_stage &desired_state)
+  {
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::wait_for_state %d", desired_state);
+    std::unique_lock<std::mutex> ulock_state (m_state_mutex);
+    m_state_cv.wait (ulock_state, [this, desired_state, &thread_ref]
```

Comment:
better to separate a lambda (or auto) variable to avoid ugly formatting

Reply:
fixed